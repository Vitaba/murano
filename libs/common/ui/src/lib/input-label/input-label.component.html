<ng-container *ngIf="!errors; else showErrors">
<div [class]="styles.container">
    <label [class]="styles.label">
      <ng-container *ngIf="data.required">
      <span [class]="styles.required"> {{data.required}}</span>
      </ng-container>
      {{ data.label }}
    </label>
  </div>
  </ng-container>
  <ng-template #showErrors>
    {{ errors | json}}
  </ng-template>