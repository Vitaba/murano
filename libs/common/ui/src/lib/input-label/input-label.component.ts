// tslint:disable: no-unsafe-any no-any
import { AfterViewInit, ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, OnChanges, SimpleChanges } from '@angular/core';
import { GuardError, InputLabelData, InputLabelStyles, isInputLabelData, isInputLabelStyles } from '@vitaba/common-utils';
import { BehaviorSubject } from 'rxjs';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-input-label',
  styleUrls: ['./input-label.component.scss'],
  templateUrl: './input-label.component.html',
})
export class InputLabelComponent implements OnChanges, AfterViewInit {
  public changes!: BehaviorSubject<SimpleChanges>;
  public errors!: GuardError;

  @Input() public data: InputLabelData = {
    label: 'Name',
  };

  @Input() public styles: InputLabelStyles = {
    container: `md:w-1/3`,
    label: `block text-gray-500 font-bold md:text-right mb-1 md:mb-0 pr-4`,
  };

  public constructor(
    private readonly _changeDetectorRef: ChangeDetectorRef,
  ) {}

  public ngOnChanges(changes: SimpleChanges): void {
    if (!this.changes) {
      this.changes = new BehaviorSubject(changes);
    }
    this.changes.next(changes);
  }

  public ngAfterViewInit(): void {
    if (!this.changes) {
      this.changes = new BehaviorSubject(null);
    }
    this.changes.subscribe((changes: SimpleChanges) => {
      if (changes.data) {
        const validation =
        this.validateInputData(changes.data.currentValue as InputLabelData);

        if (validation) {
          this.errors = { ...this.errors, ...validation };
        }
      }

      if (changes.styles) {
        const validation =
        this.validateInputLabelStyles(
          changes.styles.currentValue as InputLabelStyles);

        if (validation) {
          this.errors = { ...this.errors, ...validation };
        }
      }
      this._changeDetectorRef.detectChanges();
    });
  }

  public validateInputData(value: InputLabelData) {
    const validations = [{
      name: 'label',
      type: 'string',
    },
    ];

    const validation = isInputLabelData(value, validations);

    return !validation.valid ? validation : undefined;
  }

  public validateInputLabelStyles(value: InputLabelStyles) {
    const validations = [
      {
        name: 'container',
        type: 'string',
      },
      {
        name: 'label',
        type: 'string',
      },
    ];

    const validation = isInputLabelStyles(value, validations);

    return !validation.valid ? validation : undefined;
  }
}
