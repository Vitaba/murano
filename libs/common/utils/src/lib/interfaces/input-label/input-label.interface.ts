export interface InputLabelData {
  label: string;
  required?: string;
}

export interface InputLabelStyles {
  container: string;
  label: string;
  required?: string;
}
