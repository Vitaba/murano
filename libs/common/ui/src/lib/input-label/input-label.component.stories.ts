// tslint:disable:no-default-export object-literal-sort-keys ter-max-len

import { text } from '@storybook/addon-knobs';
import { InputLabelComponent } from './input-label.component';

export default {
  title: 'Input Label',
};

const data = {
  label: 'Name',
};

const styles = {
  container: `md:w-1/3`,
  label: `block text-gray-500 font-bold md:text-right mb-1 md:mb-0 pr-4`,
};

export const basic = () => ({
  moduleMetadata: {
    declarations: [InputLabelComponent],
    imports: [],
  },
  props: {
    data_label: text('label', data.label, 'data'),
    styles_container: text('container', styles.container, 'styles'),
    styles_label: text('label', styles.label, 'styles'),
  },
  template: `
  <vitaba-input-label
    [data]="{
    label: data_label
    }"
    [styles]="{
    container: styles_container,
    label: styles_label
    }"
    > </vitaba-input-label>
  `,
  stylesUrl: [
    'node_modules/tailwindcss/dist/base',
    'node_modules/tailwindcss/dist/components',
    'node_modules/tailwindcss/dist/utilities',
  ]
});
