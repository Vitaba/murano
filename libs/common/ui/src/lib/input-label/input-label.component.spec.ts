
// tslint:disable: no-any component-max-inline-declarations ter-max-len
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { InputLabelComponent } from './input-label.component';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-input-label-sample-implementation',
  template: `
    <vitaba-input-label [data]="data" [styles]="styles">
    </vitaba-input-label>
  `,
})
class TestHostComponent {
  public styles = {
    container: `md:w-1/3`,
    label: `block text-gray-500 font-bold md:text-right mb-1 md:mb-0 pr-4`,
  };
  public data = {
    label: 'Useful resources and assets for users.',
  };
}

describe('InputLabelComponent', () => {
  let component: TestHostComponent;
  let fixture: ComponentFixture<TestHostComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestHostComponent, InputLabelComponent],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  it('should be default', () => {
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should throw errors when data is not in right format', () => {
    component.data = {
      label: 1,
    } as any;
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
