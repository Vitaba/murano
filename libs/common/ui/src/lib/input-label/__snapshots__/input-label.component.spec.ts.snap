// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputLabelComponent should be default 1`] = `
<vitaba-input-label-sample-implementation
  data={[Function Object]}
  styles={[Function Object]}
>
  <vitaba-input-label
    ng-reflect-data="[object Object]"
    ng-reflect-styles="[object Object]"
  >
    
    
    <div
      class="md:w-1/3"
    >
      <label
        class="block text-gray-500 font-bold md:text-right mb-1 md:mb-0 pr-4"
      >
        
         Useful resources and assets for users. 
      </label>
    </div>
    
  </vitaba-input-label>
</vitaba-input-label-sample-implementation>
`;

exports[`InputLabelComponent should throw errors when data is not in right format 1`] = `
<vitaba-input-label-sample-implementation
  data={[Function Object]}
  styles={[Function Object]}
>
  <vitaba-input-label
    ng-reflect-data="[object Object]"
    ng-reflect-styles="[object Object]"
  >
    
     {
    "valid": false,
    "guard": "isInputLabelData",
    "value": {
      "label": 1
    },
    "message": "Is not a input label data",
    "errors": [
      {
        "valid": false,
        "guard": "hasProperties",
        "value": {
          "label": 1
        },
        "errors": [
          {
            "value": {
              "name": "label",
              "type": "string"
            },
            "guard": "String Validation",
            "message": "Is not a string",
            "valid": false
          }
        ]
      }
    ]
  } 
    
  </vitaba-input-label>
</vitaba-input-label-sample-implementation>
`;
